src/object/cli_obj.c

```diff
@@ -6646,6 +6649,338 @@ shard_punch_prep(struct shard_auxi_args *shard_auxi, struct dc_object *obj,
+		obj_shard_close(shard);
+
+	if (rc < 0) {
+		if (octs != NULL)
+			obj_coll_punch_cleanup(mbs, octs, node_nr + 1);
+		*length = 0;
+	} else {
+		*p_octs = octs;
+		*p_mbs = mbs;
+	}
+
+	return rc;
+}
+
+struct obj_coll_punch_cb_args {
+	crt_rpc_t	*ocpca_rpc;
+	uint32_t	*ocpca_ver;
+};
+
+static int
+obj_coll_punch_cb(tse_task_t *task, void *data)
+{
+	struct obj_coll_punch_cb_args	*cb_args = data;
+	crt_rpc_t			*rpc = cb_args->ocpca_rpc;
+	struct obj_coll_punch_in	*ocpi = crt_req_get(rpc);
+
+	if (task->dt_result == 0) {
+		task->dt_result = obj_reply_get_status(rpc);
+		*cb_args->ocpca_ver = obj_reply_map_version_get(rpc);
+	}
+
+	DL_CDEBUG(task->dt_result < 0, DLOG_ERR, DB_IO, task->dt_result,
+		  "DAOS_OBJ_RPC_COLL_PUNCH RPC %p for "DF_UOID" on %d ranks with DTX "
+		  DF_DTI" for task %p, map_ver %u/%u, flags %lx/%x\n",
+		  rpc, DP_UOID(ocpi->ocpi_oid), (int)ocpi->ocpi_tgts.ca_count,
+		  DP_DTI(&ocpi->ocpi_odm.odm_xid), task, ocpi->ocpi_map_ver, *cb_args->ocpca_ver,
+		  (unsigned long)ocpi->ocpi_api_flags, ocpi->ocpi_flags);
+
+	obj_coll_punch_cleanup(ocpi->ocpi_odm.odm_mbs, ocpi->ocpi_tgts.ca_arrays,
+			       ocpi->ocpi_tgts.ca_count);
+	crt_req_decref(rpc);
+
+	return task->dt_result;
+}
+
+static int
+dc_obj_coll_punch(tse_task_t *task, struct dc_object *obj, struct dtx_epoch *epoch, uint32_t map_ver,
+		  daos_obj_punch_t *args, struct obj_auxi_args *auxi)
+{
+	struct dc_pool			*pool = obj->cob_pool;
+	struct obj_coll_target		*octs = NULL;
+	struct dtx_memberships		*mbs = NULL;
+	crt_rpc_t			*req = NULL;
+	struct obj_coll_punch_in	*ocpi = NULL;
+	struct obj_coll_punch_cb_args	 cb_args = { 0 };
+	crt_endpoint_t			 tgt_ep = { 0 };
+	uint32_t			 length = 0;
+	int				 rc;
+	int				 i;
+
+	rc = obj_coll_punch_prep(obj, map_ver, &mbs, &octs, &length);
+	if (rc < 0)
+		goto out;
+
+	tgt_ep.ep_grp = pool->dp_sys->sy_group;
+	tgt_ep.ep_rank = octs[0].oct_rank;
+
+	rc = octs[0].oct_bitmap_sz << 3;
+	for (i = 0; i < rc; i++) {
+		if (isset(octs[0].oct_bitmap, i)) {
+			tgt_ep.ep_tag = i;
+			clrbit(octs[0].oct_bitmap, i);
+			octs[0].oct_shard_nr--;
+			break;
+		}
+	}
+
+	D_ASSERT(i < rc);
+
+	rc = obj_req_create(daos_task2ctx(task), &tgt_ep, DAOS_OBJ_RPC_COLL_PUNCH, &req);
+	if (rc != 0)
+		goto out;
+
+	ocpi = crt_req_get(req);
+	D_ASSERT(ocpi != NULL);
+
+	ocpi->ocpi_odm.odm_mbs = mbs;
+	uuid_copy(ocpi->ocpi_pool_uuid, pool->dp_pool);
+	uuid_copy(ocpi->ocpi_co_hdl, obj->cob_co->dc_cont_hdl);
+	uuid_copy(ocpi->ocpi_co_uuid, obj->cob_co->dc_uuid);
+	daos_dc_obj2id(obj, &ocpi->ocpi_oid);
+	ocpi->ocpi_oid.id_shard = octs[0].oct_shards[i];
+	ocpi->ocpi_epoch = epoch->oe_value;
+	ocpi->ocpi_api_flags = args->flags;
+	ocpi->ocpi_map_ver = map_ver;
+
+	/* Always synchronously commit the DTX for object collective punch. */
+	ocpi->ocpi_flags = ORF_LEADER | ORF_DTX_SYNC;
+	if (auxi->io_retry)
+		ocpi->ocpi_flags |= ORF_RESEND;
+	else
+		daos_dti_gen(&ocpi->ocpi_odm.odm_xid, false);
+	if (obj_is_ec(obj))
+		ocpi->ocpi_flags |= ORF_EC;
```

Comment:
seems ORF_EC is not really needed for coll punch at server side

Reply:
Currently, it is yes. Packing the EC flag is just for potential special handling for EC object in future.